src/parser/binary_parser.ml

```diff
@@ -842,13 +867,16 @@ let read_code types input =
   let+ code, input = read_expr types [] input in
   ((locals, code), input)
 
-let read_data types input =
+let read_data types memories input =
   let* i, input = read_U32 input in
   match i with
   | 0 ->
     let* expr, input = read_expr types [] input in
-    let+ bytes, input = read_bytes input in
+    let* bytes, input = read_bytes input in
     let init = string_of_char_list bytes in
+    let+ () =
+      if List.is_empty memories then Error (`Unknown_memory 0) else Ok ()
```

Comment:
Could you add a `TODO: this should be removed once we do proper validation of binary modules ?` ?